File: polyscan/testdata/cpp/sample.cpp
#include <cstdio>
#include <utility>

// Adds the positive values and reports how many there were.
std::pair<int, int> sumPositive(const int* values, int n) {
  int total = 0;
  int count = 0;
  for (int i = 0; i < n; i++) {
    if (values[i] > 0) {
      total += values[i];
      count++;
    }
  }
  std::printf("%d positive values\n", count);
  return {total, count};
}

// sumPositive with the names and the comparison changed.
std::pair<int, int> sumNegative(const int* numbers, int n) {
  int sum = 0;
  int seen = 0;
  for (int i = 0; i < n; i++) {
    if (numbers[i] < 0) {
      sum += numbers[i];
      seen++;
    }
  }
  std::printf("%d negative values\n", seen);
  return {sum, seen};
}

class Server {
 public:
  // Two loops, two switch cases, an if, its &&, a ternary and a catch are
  // eight decision points, so the complexity is 9.
  int handle(const int* items, int n, int mode) const {
    int acc = 0;
    for (int i = 0; i < 10; i++) {
      acc += i;
    }
    while (acc > 100) {
      acc--;
    }
    switch (mode) {
      case 1:
        acc += 10;
        break;
      case 2:
        acc += 20;
        break;
      default:
        break;
    }
    if (acc > 5 && n > 0) {
      acc += items[0];
    }
    try {
      acc += n > 0 ? n : -n;
    } catch (...) {
      acc = 0;
    }
    return acc;
  }
};

File: polyscan/testdata/rust/sample.rs
/// Adds the positive values and reports how many there were.
pub fn sum_positive(values: &[i32]) -> (i32, usize) {
    let mut total = 0;
    let mut count = 0;
    for value in values {
        if *value > 0 {
            total += value;
            count += 1;
        }
    }
    println!("{} positive values", count);
    (total, count)
}

/// sum_positive with the names and the comparison changed.
pub fn sum_negative(numbers: &[i32]) -> (i32, usize) {
    let mut sum = 0;
    let mut seen = 0;
    for number in numbers {
        if *number < 0 {
            sum += number;
            seen += 1;
        }
    }
    println!("{} negative values", seen);
    (sum, seen)
}

pub struct Server;

impl Server {
    /// Two loops, a match with three arms, an if with && and a ? make 8.
    pub fn handle(&self, items: &[i32], n: i32, r: Result<i32, ()>) -> Result<i32, ()> {
        let mut acc = 0;
        for i in 0..10 {
            acc += i;
        }
        while acc > 100 {
            acc -= 1;
        }
        let bonus = match n {
            1 => 10,
            2 | 3 => 20,
            _ => 0,
        };
        if acc > 5 && n > 0 {
            acc += bonus;
        }
        let v = r?;
        Ok(acc + v + items.len() as i32)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Another copy of sum_positive, which test code excludes from clones.
    #[test]
    fn sums_positive_values() {
        let values = [1, -2, 3];
        let mut total = 0;
        let mut count = 0;
        for value in values.iter() {
            if *value > 0 {
                total += value;
                count += 1;
            }
        }
        assert_eq!((total, count), (4, 2));
        assert_eq!(sum_positive(&values), (4, 2));
    }
}
